src/lib/cltable.lua

```diff
@@ -0,0 +1,66 @@
+module(..., package.seeall)
+
+local ffi = require("ffi")
+local ctable = require("lib.ctable")
+
+function build(keys, values)
+   return setmetatable({ keys = keys, values = values },
+                       {__index=get, __newindex=set})
+end
+
+function new(params)
+   local ctable_params = {}
+   for k,v in _G.pairs(params) do ctable_params[k] = v end
+   assert(not ctable_params.value_type)
+   ctable_params.value_type = ffi.typeof('uint32_t')
+   return build(ctable.new(ctable_params), {})
+end
+
+function get(cltable, key)
+   local entry = cltable.keys:lookup_ptr(key)
+   if not entry then return nil end
+   return cltable.values[entry.value]
+end
+
+function set(cltable, key, value)
+   local entry = cltable.keys:lookup_ptr(key)
+   if entry then
+      cltable.values[entry.value] = value
+      if value ~= nil then
+         cltable.keys:remove_ptr(entry)
+         -- FIXME: Leaking the slot in the values array.
```

Comment:
Wait, shouldn’t this be `value == nil`, as in “if value is nil, delete the entry and set the corresponding key in the backing table to nil, too”.